cmd/agent_auditor.go

```diff
@@ -14,74 +16,177 @@
-			if err != nil {
-				log.Fatalf("Failed to join the cluster: %v", err)
-			}
-			log.Debugf("Number of nodes contacted: %d (%v)", contacted, config.StartJoin)
+func configAuditor() context.Context {
+	conf := newAuditorConfig()
+	err := gpflag.ParseTo(conf, agentAuditorCmd.PersistentFlags())
+	if err != nil {
+		log.Fatalf("err: %v", err)
+	}
 
-			defer agent.Shutdown()
-			util.AwaitTermSignal(agent.Leave)
-		},
+	ctx := context.WithValue(agentCtx, k("auditor.config"), conf)
+
+	return ctx
+}
+
+func runAgentAuditor(cmd *cobra.Command, args []string) error {
+	agentConfig := agentAuditorCtx.Value(k("agent.config")).(*gossip.Config)
+	conf := agentAuditorCtx.Value(k("auditor.config")).(*auditorConfig)
+
+	log.SetLogger("auditor", agentConfig.Log)
+
+	notifier := gossip.NewSimpleNotifierFromConfig(conf.Notifier)
+	qed, err := client.NewHTTPClientFromConfig(conf.Qed)
+	if err != nil {
+		return err
+	}
+	tm := gossip.NewSimpleTasksManagerFromConfig(conf.Tasks)
+	store := gossip.NewRestSnapshotStoreFromConfig(conf.Store)
+
+	agent, err := gossip.NewDefaultAgent(agentConfig, qed, store, tm, notifier)
+	if err != nil {
+		return err
 	}
 
-	f := cmd.Flags()
-	f.StringSliceVarP(&auditorConfig.QEDUrls, "qedUrls", "", []string{}, "Comma-delimited list of QED servers ([host]:port), through which an auditor can make queries")
-	f.StringSliceVarP(&auditorConfig.PubUrls, "pubUrls", "", []string{}, "Comma-delimited list of store servers ([host]:port), through which an auditor can make queries")
-	f.StringSliceVarP(&auditorConfig.AlertsUrls, "alertsUrls", "", []string{}, "Comma-delimited list of alerts servers ([host]:port), through which an auditor can make queries")
-	// Lookups
-	v.BindPFlag("agent.server_urls", f.Lookup("qedUrls"))
-	v.BindPFlag("agent.snapshots_store_urls", f.Lookup("pubUrls"))
-	v.BindPFlag("agent.alerts_urls", f.Lookup("alertsUrls"))
-	return cmd
+	bp := gossip.NewBatchProcessor(agent, []gossip.TaskFactory{gossip.PrinterFactory{}, membershipFactory{}})
+	agent.In.Subscribe(gossip.BatchMessageType, bp, 255)
+	defer bp.Stop()
+
+	agent.Start()
+	util.AwaitTermSignal(agent.Shutdown)
+	return nil
+}
+
+type membershipFactory struct{}
+
+func (m membershipFactory) Metrics() []prometheus.Collector {
+	return []prometheus.Collector{
+		QedAuditorInstancesCount,
+		QedAuditorBatchesProcessSeconds,
+		QedAuditorBatchesReceivedTotal,
+		QedAuditorGetMembershipProofErrTotal,
+	}
+}
+
+func (i membershipFactory) New(ctx context.Context) gossip.Task {
+	a := ctx.Value("agent").(*gossip.Agent)
+	b := ctx.Value("batch").(*protocol.BatchSnapshots)
+
+	s := b.Snapshots[0]
+
+	return func() error {
+		timer := prometheus.NewTimer(QedAuditorBatchesProcessSeconds)
+		defer timer.ObserveDuration()
+
+		proof, err := a.Qed.MembershipDigest(s.Snapshot.EventDigest, s.Snapshot.Version)
+		if err != nil {
+			log.Infof("Auditor is unable to get membership proof from QED server: %v", err)
+
+			switch fmt.Sprintf("%T", err) {
+			case "*errors.errorString":
```

Comment:
Why are you checking this?

Reply:
To check whether the membership API is working properly or not. Is of little value probably, just monitoring it is available and working.